### PR TITLE
CPU management for HQ

### PR DESCRIPTION
**Pull request by spirali**

### CODE REVIEW

File: src/client/commands/submit.rs [hunk 1]
```diff
@@ -6,11 +6,26 @@ use tako::messages::common::ProgramDefinition;
...
+use crate::client::resources::parse_cpu_request;
 use crate::common::arraydef::ArrayDef;
 use crate::rpc_call;
 use crate::transfer::connection::ClientConnection;
 use crate::transfer::messages::{FromClientMessage, JobType, SubmitRequest, ToClientMessage};
 use anyhow::anyhow;
+use std::str::FromStr;
+use tako::common::resources::{CpuRequest, ResourceRequest};
+
+// I am wrapping CpuRequest + implementing FromStr as I am not able to provide
```
Comment (Kobzol) [inline src/client/commands/submit.rs:18]: I also didn't find any other way currently.

File: src/client/resources.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            tuple((
+                p_uint,
+                opt(preceded(
+                    multispace1,
+                    alt((tag("compact!"), tag("compact"), tag("scatter"))),
+                )),
+            )),
+            |(count, policy)| {
+                if count == 0 {
+                    return Err(nom::Err::Error("xxx"));
```
Comment (Kobzol) [inline src/client/resources.rs:23]: xxx? :)

File: src/worker/start.rs [hunk 3]
```diff
@@ -57,6 +61,21 @@ pub struct WorkerStartOpts {
     manager: ManagerOpts,
 }
 
+fn launcher_setup(task: &Task, def: LauncherDefinition) -> tako::Result<ProgramDefinition> {
+    let allocation = task.resource_allocation().unwrap();
```
Comment (Kobzol) [inline src/worker/start.rs:65]: `?`

**Review by Kobzol — APPROVED** [review]

File: src/client/resources.rs (hunk 2 above)
Comment (spirali) [reply src/client/resources.rs:23]: Oops :)

File: src/worker/start.rs (hunk 3 above)
Comment (spirali) [reply src/worker/start.rs:65]: launcher setup should be executed only on a running task and such that it has an allocation. It is a fatal internal error if allocation is not there, so I decided for unwrap.

Comment (Kobzol) [reply src/worker/start.rs:65]: Ok. I suggest to always use `expect` in such cases so that we have documentation about why it should not fail.

**Pull request merged**